File: Java/LeetCode/Count Pairs Of Similar Strings.java
// https://leetcode.com/problems/count-pairs-of-similar-strings

class Solution {

    public int similarPairs(String[] words) {
        int similarPairCount = 0;

        for (int i = 0; i < words.length; ++i) {
            for (int j = i+1; j < words.length; ++j) {
                Set<Character> firstWordChars = new HashSet<>();
                Set<Character> secondWordChars = new HashSet<>();
                for (int k = 0; k < words[i].length(); ++k) {
                    firstWordChars.add( words[i].charAt(k) );
                }
                for (int m = 0; m < words[j].length(); ++m) {
                    secondWordChars.add( words[j].charAt(m) );
                }
                if (firstWordChars.equals(secondWordChars)) {
                    similarPairCount++;
                }
            }
        }

        return similarPairCount;
    }

}

// Alternative solution
/*
class Solution {

    public int similarPairs(String[] words) {
        int similarPairCount = 0;

        for (int i = 0; i < words.length; ++i) {
            for (int j = i+1; j < words.length; ++j) {
                Set<Character> s1 = new HashSet<Character>(
                    Arrays.asList(words[i].chars().mapToObj(c -> (char)c).toArray(Character[]::new))
                );
                Set<Character> s2 = new HashSet<>(
                    Arrays.asList(words[j].chars().mapToObj(c -> (char)c).toArray(Character[]::new))
                );
                if (s1.equals(s2)) {
                    similarPairCount++;
                }
            }
        }

        return similarPairCount;
    }
}
*/

File: Kotlin/LeetCode/Count Pairs Of Similar Strings.kt
// https://leetcode.com/problems/count-pairs-of-similar-strings/submissions/1373367005

class Solution {
  
    fun similarPairs(words: Array<String>): Int {
        var similarPairCount = 0

        for (i in words.indices) {
            for (j in i + 1 until words.size) {
                val s1 = words[i].toSet()
                val s2 = words[j].toSet()
                if (s1 == s2) {
                    similarPairCount++
                }
            }
        }

        return similarPairCount
    }
    
}

// Alternative solution
/*
class Solution {
  
    fun similarPairs(words: Array<String>): Int {
        var similarPairCount = 0

        for (i in 0 until words.size) {
            for (j in i+1 until words.size) {
                val firstWordChars = mutableSetOf<Char>()
                val secondWordChars = mutableSetOf<Char>()
                for (k in 0 until words[i].length) {
                    firstWordChars.add(words[i][k])
                }
                for (m in 0 until words[j].length) {
                    secondWordChars.add(words[j][m])
                }
                if (firstWordChars.equals(secondWordChars)) {
                    similarPairCount++
                }
            }
        }

        return similarPairCount
    }
    
}
*/
